models/settings_helpers.go

```diff
@@ -44,39 +48,148 @@ func GetSettings(q reform.DBTX) (*Settings, error) {
+func UpdateSettings(q reform.DBTX, params ChangeSettingsParams) (*Settings, error) {
+	err := ValidateSettings(params)
+	if err != nil {
+		return nil, err
+	}
+	settings, err := GetSettings(q)
+	if err != nil {
+		return nil, err
+	}
+	if params.DisableTelemetry {
+		settings.Telemetry.Disabled = true
+	}
+	if params.EnableTelemetry {
+		settings.Telemetry.Disabled = false
+	}
+	if params.MetricsResolutions.LR != 0 {
+		settings.MetricsResolutions.LR = params.MetricsResolutions.LR
+	}
+	if params.MetricsResolutions.MR != 0 {
+		settings.MetricsResolutions.MR = params.MetricsResolutions.MR
+	}
+	if params.MetricsResolutions.HR != 0 {
+		settings.MetricsResolutions.HR = params.MetricsResolutions.HR
+	}
+	if params.DataRetention != 0 {
+		settings.DataRetention = params.DataRetention
+	}
 
-	for _, pair := range []struct {
-		dur  time.Duration
-		name string
+	if len(params.AWSPartitions) != 0 {
+		settings.AWSPartitions = deduplicateAWSPartitions(params.AWSPartitions)
+	}
+	if params.SSHKey != "" {
+		settings.SSHKey = params.SSHKey
+	}
+	if params.AlertManagerURL != "" {
+		settings.AlertManagerURL = params.AlertManagerURL
+	}
+	if params.RemoveAlertManagerURL {
+		settings.AlertManagerURL = ""
+	}
+
+	err = SaveSettings(q, settings)
+	if err != nil {
+		return nil, err
+	}
+	return settings, nil
+}
+
+// ValidateSettings validates settings changes.
+func ValidateSettings(params ChangeSettingsParams) error {
+	if params.EnableTelemetry && params.DisableTelemetry {
+		return status.Error(codes.InvalidArgument, "Both enable_telemetry and disable_telemetry are present.")
+	}
+
+	checkCases := []struct {
+		dur       time.Duration
+		fieldName string
+		validator func(time.Duration) (time.Duration, error)
 	}{
-		{dur: s.MetricsResolutions.HR, name: "hr"},
-		{dur: s.MetricsResolutions.MR, name: "mr"},
-		{dur: s.MetricsResolutions.LR, name: "lr"},
-	} {
-		if pair.dur < time.Second {
-			return status.Error(codes.InvalidArgument, pair.name+": minimal resolution is 1s")
+		{params.MetricsResolutions.HR, "hr", validators.ValidateMetricResolution},
+		{params.MetricsResolutions.MR, "mr", validators.ValidateMetricResolution},
+		{params.MetricsResolutions.LR, "lr", validators.ValidateMetricResolution},
+	}
+	for _, v := range checkCases {
+		if v.dur == 0 {
+			continue
 		}
-		if pair.dur.Truncate(time.Second) != pair.dur {
-			return status.Error(codes.InvalidArgument, pair.name+": should be a natural number of seconds")
+
+		if _, err := v.validator(v.dur); err != nil {
+			switch err.(type) {
+			case validators.DurationNotAllowedError:
+				return status.Error(codes.InvalidArgument, fmt.Sprintf("%s: should be a natural number of seconds", v.fieldName))
+			case validators.MinDurationError:
+				return status.Error(codes.InvalidArgument, fmt.Sprintf("%s: minimal resolution is 1s", v.fieldName))
+			default:
+				return status.Error(codes.InvalidArgument, fmt.Sprintf("%s: unknown error for", v.fieldName))
```

Comment:
In cases like that we should just return `err` – non-gRPC errors are properly handled by interceptor.